File: scripts/fetch-rates.js
#!/usr/bin/env node
/**
 * Mortgage Rate Fetcher v3
 * 
 * Extracts structured rate data from:
 * - Bankrate's __NEXT_DATA__ JSON (national lenders)
 * - PatelCo Credit Union (local CU rates)
 * - Star One Credit Union (local CU rates)
 * - National averages (from page text)
 * - ARM rates from dedicated ARM page
 * 
 * Outputs:
 *   data/rates-YYYY-MM-DD.json  (full historical record)
 *   site/data/latest.json       (for public website)
 *   data/last-change-report.txt (if rates changed vs yesterday)
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..');
const DATA_DIR = path.join(ROOT, 'data');
const SITE_DATA_DIR = path.join(ROOT, 'site', 'data');

[DATA_DIR, SITE_DATA_DIR].forEach(d => {
  if (!fs.existsSync(d)) fs.mkdirSync(d, { recursive: true });
});

// Your loan profile
const MY_PROFILE = {
  currentRate: 5.8,
  loanType: "5/1 ARM",
  location: "Milpitas, CA 95035",
  propertyType: "Single-family detached",
};

function fetch(url) {
  return new Promise((resolve, reject) => {
    const req = https.request(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml',
        'Accept-Language': 'en-US,en;q=0.9',
      }
    }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return fetch(res.headers.location).then(resolve).catch(reject);
      }
      let body = '';
      res.on('data', c => body += c);
      res.on('end', () => resolve({ body, statusCode: res.statusCode }));
    });
    req.on('error', reject);
    req.setTimeout(20000, () => { req.destroy(); reject(new Error('timeout: ' + url)); });
    req.end();
  });
}

function extractNextData(body) {
  const m = body.match(/__NEXT_DATA__"[^>]*>([^<]+)</i);
  if (!m) throw new Error('No __NEXT_DATA__ found');
  return JSON.parse(m[1]);
}

function parseProducts(data) {
  const query = data.props?.pageProps?.dehydratedState?.queries?.[0];
  if (!query) return { lenders: [], queryKey: null };
  
  const d = query.state?.data;
  if (!d) return { lenders: [], queryKey: query.queryKey };
  
  const institutions = d.institutions || {};
  const terms = d.terms || {};
  
  const lenders = (d.products || []).map(p => {
    const inst = institutions[p.institutionId] || {};
    const off = p.offering || {};
    const dims = p.dimensions || {};
    
    return {
      lender: inst.name || inst.displayName || 'Unknown',
      lenderId: p.institutionId,
      logo: inst.logo || null,
      product: p.name || p.displayName || '',
      productTypeId: p.productTypeId,
      rate: off.rate || null,
      apr: off.apr || null,
      points: off.points || null,
      monthlyPayment: off.estimatedPayment || off.principalAndInterestPayment || null,
      upFrontCosts: off.upFrontCosts || null,
      fiveYearCost: off.fiveYearCost || null,
      eightYearCost: off.eightYearCost || null,
      interestType: dims.interestType || null,
      loanSize: dims.size || null,
      isFHA: dims.isFha || false,
      isVA: dims.isVa || false,
      lockDays: p.eligibility?.lockDays || null,
    };
  }).filter(l => l.rate !== null);
  
  return { lenders, queryKey: query.queryKey };
}

// Parse national averages from page text
// Strips HTML tags first for reliable regex matching
function parseNationalAverages(body) {
  // Strip HTML tags and normalize whitespace
  const text = body.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ');
  const rates = {};
  
  // National averages: "30-year fixed ... 6.58%"
  const m30 = text.match(/30.year.fixed[^]{0,300}?(\d+\.\d{2,3})%/i);
  if (m30) rates['30_year_fixed'] = parseFloat(m30[1]);
  
  const m15 = text.match(/15.year.fixed[^]{0,300}?(\d+\.\d{2,3})%/i);
  if (m15) rates['15_year_fixed'] = parseFloat(m15[1]);
  
  // ARM national average: "the national average 5/1 ARM APR is 6.18%"
  const armMatch = text.match(/national average 5.1.ARM APR is\s+([\d.]+)%/i);
  if (armMatch) rates['5_1_arm_apr'] = parseFloat(armMatch[1]);
  
  const arm101Match = text.match(/10.1.ARM APR is\s+([\d.]+)%/i);
  if (arm101Match) rates['10_1_arm_apr'] = parseFloat(arm101Match[1]);
  
  // Weekly trend block: "5/1 ARM 6.20% 15 year fixed 5.89% 30 year fixed 6.52%"
  const weekly = text.match(/5.1.ARM\s+([\d.]+)%\s+15.year.fixed\s+([\d.]+)%\s+30.year.fixed\s+([\d.]+)%/i);
  if (weekly) {
    rates.weekly_trend = {
      arm_5_1: parseFloat(weekly[1]),
      fixed_15: parseFloat(weekly[2]),
      fixed_30: parseFloat(weekly[3]),
    };
  }
  
  // Top daily offer
  const topOffer = text.match(/top offers on Bankrate:\s+([\d.]+)%/i);
  if (topOffer) rates.top_daily_offer = parseFloat(topOffer[1]);
  
  return rates;
}

// Parse PatelCo Credit Union rates from their mortgage page HTML
function parsePatelCo(body) {
  const text = body.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ');
  const fixed = [];
  const arm = [];

  // Fixed-rate products: "30-Year Fixed $50,000 to $832,750 6.625% 6.710%"
  const fixedRegex = /(\d+)-Year Fixed(?:\s+(?:High Balance|Jumbo))?\s+\$[\d,]+\s+to\s+\$[\d,]+\s+(\d+\.\d{2,4})%\s+(\d+\.\d{2,4})%/gi;
  const seen = new Set();
  let m;
  while ((m = fixedRegex.exec(text)) !== null) {
    const key = `${m[1]}yr-${m[2]}`;
    if (seen.has(key)) continue;
    seen.add(key);
    fixed.push({
      term: `${m[1]}-Year Fixed`,
      rate: parseFloat(m[2]),
      apr: parseFloat(m[3]),
    });
  }

  // ARM products: "5/1 30-Year Adjustable 6.125% 6.279%"
  const armRegex = /(\d+\/\d+)\s+30-Year Adjustable(?:\s+Jumbo)?\s+\$[\d,]+\s+to\s+\$[\d,]+\s+(\d+\.\d{2,4})%\s+(\d+\.\d{2,4})%/gi;
  const seenArm = new Set();
  while ((m = armRegex.exec(text)) !== null) {
    const key = `${m[1]}-${m[2]}`;
    if (seenArm.has(key)) continue;
    seenArm.add(key);
    arm.push({
      term: `${m[1]} ARM`,
      rate: parseFloat(m[2]),
      apr: parseFloat(m[3]),
    });
  }

  return {
    name: 'PatelCo Credit Union',
    url: 'https://www.patelco.org/credit-cards-and-loans/home-loans/mortgage',
    fixed,
    arm,
  };
}

// Parse Star One Credit Union rates from their mortgage page HTML
function parseStarOne(body) {
  const text = body.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ');
  const fixed = [];
  const arm = [];

  // Fixed rates: "N-year mortgage at X.XXX% (Y.YYY% APR"
  const fixedRegex = /(\d+)-year mortgage at (\d+\.\d{2,4})%\s*\((\d+\.\d{2,4})%\s*APR/gi;
  const seen = new Set();
  let m;
  while ((m = fixedRegex.exec(text)) !== null) {
    const key = `${m[1]}yr`;
    if (seen.has(key)) continue;
    seen.add(key);
    fixed.push({
      term: `${m[1]}-Year Fixed`,
      rate: parseFloat(m[2]),
      apr: parseFloat(m[3]),
    });
  }

  // ARM rates: "N-Year Fixed-to-Adjustable-Rate Mortgage ... is X.XXX% (Y.YYY% APR for conforming"
  const armRegex = /(\d+)-Year Fixed-to-Adjustable-Rate Mortgage.*?is (\d+\.\d{2,4})%\s*\((\d+\.\d{2,4})%\s*APR for conforming/gi;
  const seenArm = new Set();
  while ((m = armRegex.exec(text)) !== null) {
    const key = `${m[1]}yr`;
    if (seenArm.has(key)) continue;
    seenArm.add(key);
    arm.push({
      term: `${m[1]}/1 ARM`,
      rate: parseFloat(m[2]),
      apr: parseFloat(m[3]),
    });
  }

  return {
    name: 'Star One Credit Union',
    url: 'https://www.starone.org/rates/mortgage-rates',
    fixed,
    arm,
  };
}

async function main() {
  const today = new Date().toISOString().split('T')[0];
  const now = new Date().toISOString();
  console.log(`[${now}] Fetching mortgage rates for ${today}...`);

  const result = {
    date: today,
    fetchedAt: now,
    profile: MY_PROFILE,
    nationalAverages: {},
    lenderOffers: [],
    armRates: {},
    summary: {},
  };

  // === Fetch Bankrate main page ===
  try {
    console.log('→ Bankrate main rates page...');
    const res = await fetch('https://www.bankrate.com/mortgages/mortgage-rates/');
    const nextData = extractNextData(res.body);
    const { lenders, queryKey } = parseProducts(nextData);
    result.lenderOffers = lenders;
    result.queryParams = queryKey;
    console.log(`  ✓ ${lenders.length} lender offers extracted`);

    const natAvg = parseNationalAverages(res.body);
    result.nationalAverages = { ...result.nationalAverages, ...natAvg };
    console.log('  ✓ National averages:', JSON.stringify(natAvg));
  } catch (e) {
    console.error('  ✗ Bankrate main failed:', e.message);
    result.errors = result.errors || [];
    result.errors.push({ source: 'bankrate_main', error: e.message });
  }

  // === Fetch Bankrate ARM page ===
  try {
    console.log('→ Bankrate ARM rates page...');
    const res = await fetch('https://www.bankrate.com/mortgages/5-1-arm-rates/');
    const armData = parseNationalAverages(res.body);
    result.armRates = armData;
    
    // Try to get ARM lender offers too
    try {
      const nextData = extractNextData(res.body);
      const { lenders: armLenders } = parseProducts(nextData);
      if (armLenders.length) {
        result.armLenderOffers = armLenders;
        console.log(`  ✓ ${armLenders.length} ARM lender offers extracted`);
      }
    } catch (e) {
      // ARM page might not have structured data
    }
    
    console.log('  ✓ ARM rates:', JSON.stringify(armData));
  } catch (e) {
    console.error('  ✗ Bankrate ARM failed:', e.message);
  }

  // === Fetch PatelCo Credit Union rates ===
  try {
    console.log('→ PatelCo Credit Union rates...');
    const res = await fetch('https://www.patelco.org/credit-cards-and-loans/home-loans/mortgage');
    result.creditUnions = result.creditUnions || {};
    result.creditUnions.patelco = parsePatelCo(res.body);
    console.log(`  ✓ PatelCo: ${result.creditUnions.patelco.fixed.length} fixed, ${result.creditUnions.patelco.arm.length} ARM rates`);
  } catch (e) {
    console.error('  ✗ PatelCo failed:', e.message);
    result.errors = result.errors || [];
    result.errors.push({ source: 'patelco', error: e.message });
  }

  // === Fetch Star One Credit Union rates ===
  try {
    console.log('→ Star One Credit Union rates...');
    const res = await fetch('https://www.starone.org/rates/mortgage-rates');
    result.creditUnions = result.creditUnions || {};
    result.creditUnions.starone = parseStarOne(res.body);
    console.log(`  ✓ Star One: ${result.creditUnions.starone.fixed.length} fixed, ${result.creditUnions.starone.arm.length} ARM rates`);
  } catch (e) {
    console.error('  ✗ Star One failed:', e.message);
    result.errors = result.errors || [];
    result.errors.push({ source: 'starone', error: e.message });
  }

  // === Build summary ===
  // Merge national averages with ARM-specific data
  const na = { ...(result.armRates || {}), ...(result.nationalAverages || {}) };
  const wt = na.weekly_trend || {};
  
  result.summary = {
    '30_year_fixed': wt.fixed_30 || na['30_year_fixed'] || null,
    '15_year_fixed': wt.fixed_15 || na['15_year_fixed'] || null,
    '5_1_arm': wt.arm_5_1 || na['5_1_arm_apr'] || null,
    '10_1_arm': na['10_1_arm_apr'] || null,
    'top_daily_30yr_offer': na.top_daily_offer || null,
    'best_30yr_lender_rate': result.lenderOffers.length
      ? Math.min(...result.lenderOffers.filter(l => l.product?.includes('30')).map(l => l.rate))
      : null,
    'my_rate': MY_PROFILE.currentRate,
    'my_loan_type': MY_PROFILE.loanType,
  };

  // Compare to your rate
  const marketARM = result.summary['5_1_arm'];
  if (marketARM) {
    const diff = marketARM - MY_PROFILE.currentRate;
    result.summary.vs_my_rate = {
      market_5_1_arm: marketARM,
      my_rate: MY_PROFILE.currentRate,
      difference: parseFloat(diff.toFixed(3)),
      i_save_or_lose_per_month: null, // would need loan amount
      assessment: diff > 0.5 ? '🌟 Your rate is EXCELLENT — well below market'
                 : diff > 0.25 ? '✅ Your rate is BETTER than market'
                 : diff > 0.05 ? '🟢 Your rate is slightly better than market'
                 : diff > -0.05 ? '🟡 Market is about equal to your rate'
                 : diff > -0.25 ? '🟠 Market is slightly better — watch for refinance'
                 : '🔴 Market is BETTER — consider refinancing',
    };
  }

  // === Save files ===
  const dayFile = path.join(DATA_DIR, `rates-${today}.json`);
  fs.writeFileSync(dayFile, JSON.stringify(result, null, 2));
  console.log(`\nSaved: ${dayFile}`);

  // Latest for website
  const latestFile = path.join(SITE_DATA_DIR, 'latest.json');
  fs.writeFileSync(latestFile, JSON.stringify(result, null, 2));
  console.log(`Saved: ${latestFile}`);

  // === Detect changes vs yesterday ===
  const yesterdayDate = new Date(Date.now() - 86400000).toISOString().split('T')[0];
  const yFile = path.join(DATA_DIR, `rates-${yesterdayDate}.json`);
  const changes = [];

  if (fs.existsSync(yFile)) {
    const yData = JSON.parse(fs.readFileSync(yFile, 'utf8'));
    for (const key of ['30_year_fixed', '15_year_fixed', '5_1_arm', '10_1_arm', 'top_daily_30yr_offer', 'best_30yr_lender_rate']) {
      const t = result.summary[key];
      const y = yData.summary?.[key];
      if (t != null && y != null && Math.abs(t - y) >= 0.005) {
        const arrow = t > y ? '📈' : '📉';
        const delta = (t - y).toFixed(3);
        changes.push(`${arrow} ${key.replace(/_/g, ' ')}: ${y.toFixed(3)}% → ${t.toFixed(3)}% (${delta > 0 ? '+' : ''}${delta}%)`);
      }
    }
  }

  // Print summary
  console.log('\n' + '='.repeat(60));
  console.log('📊 MORTGAGE RATE DASHBOARD — ' + today);
  console.log('='.repeat(60));
  
  const s = result.summary;
  const fmt = v => v != null ? v.toFixed(2) + '%' : 'N/A';
  
  console.log(`30-Year Fixed:      ${fmt(s['30_year_fixed'])}`);
  console.log(`15-Year Fixed:      ${fmt(s['15_year_fixed'])}`);
  console.log(`5/1 ARM:            ${fmt(s['5_1_arm'])}`);
  console.log(`10/1 ARM:           ${fmt(s['10_1_arm'])}`);
  console.log(`Top 30yr Daily:     ${fmt(s['top_daily_30yr_offer'])}`);
  console.log(`Best Lender 30yr:   ${fmt(s['best_30yr_lender_rate'])}`);
  console.log('-'.repeat(60));
  console.log(`Your Rate (5/1 ARM): ${MY_PROFILE.currentRate}%`);
  
  if (s.vs_my_rate) {
    console.log(`Market 5/1 ARM:     ${s.vs_my_rate.market_5_1_arm.toFixed(2)}%`);
    console.log(`Difference:         ${s.vs_my_rate.difference > 0 ? '+' : ''}${s.vs_my_rate.difference}%`);
    console.log(`Status:             ${s.vs_my_rate.assessment}`);
  }

  if (result.lenderOffers.length) {
    console.log('\n--- Top Lender Offers ---');
    const byRate = [...result.lenderOffers].sort((a, b) => (a.rate || 99) - (b.rate || 99));
    byRate.slice(0, 5).forEach((l, i) => {
      console.log(`  ${i+1}. ${l.lender} — ${l.product}: ${l.rate}% (APR ${l.apr}%, ${l.points} pts, $${Math.round(l.monthlyPayment || 0)}/mo)`);
    });
  }

  // Print credit union rates
  if (result.creditUnions) {
    console.log('\n--- Local Credit Union Rates ---');
    for (const [key, cu] of Object.entries(result.creditUnions)) {
      console.log(`\n  ${cu.name}:`);
      for (const f of cu.fixed) {
        const marker = f.term === '30-Year Fixed' ? ' ←' : '';
        console.log(`    ${f.term}: ${f.rate}% (APR ${f.apr}%)${marker}`);
      }
      for (const a of cu.arm) {
        const marker = a.term === '5/1 ARM' || a.term === '5/1 ARM' ? ' ←' : '';
        console.log(`    ${a.term}: ${a.rate}% (APR ${a.apr}%)${marker}`);
      }
    }
  }

  if (changes.length) {
    console.log('\n🔔 RATE CHANGES SINCE YESTERDAY:');
    changes.forEach(c => console.log('  ' + c));
    
    const report = `📊 MORTGAGE RATE CHANGES — ${today}\n\n${changes.join('\n')}\n\n` +
      `Your ${MY_PROFILE.loanType}: ${MY_PROFILE.currentRate}%\n` +
      (s.vs_my_rate ? `Market 5/1 ARM: ${s.vs_my_rate.market_5_1_arm.toFixed(2)}% — ${s.vs_my_rate.assessment}` : '');
    fs.writeFileSync(path.join(DATA_DIR, 'last-change-report.txt'), report);
    console.log('\n⚠️ Change alert saved.');
  } else {
    const rptFile = path.join(DATA_DIR, 'last-change-report.txt');
    if (fs.existsSync(rptFile)) fs.unlinkSync(rptFile);
    console.log('\n✓ No significant changes since yesterday.');
  }

  console.log('='.repeat(60));
  return result;
}

main().catch(err => {
  console.error('FATAL:', err);
  process.exit(1);
});

File: fetch_rates.py
#!/usr/bin/env python3
"""
Mortgage Rate Fetcher v4 — Direct lender sources
  Credit Unions: PatelCo, Star One, Golden 1
  Banks: Bank of America, Wells Fargo, Chase
  National: Freddie Mac PMMS

Output:
  data/rates-YYYY-MM-DD.json      — Daily snapshot
  site/data/latest.json            — Latest for website
  Supabase mortgage_rates table    — Historical tracking
"""

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

import httpx
from httpx import Client

ROOT = Path(__file__).parent
DATA_DIR = ROOT / "data"
SITE_DATA_DIR = ROOT / "site" / "data"
SCRAPERS_DIR = ROOT / "scrapers"

DATA_DIR.mkdir(parents=True, exist_ok=True)
SITE_DATA_DIR.mkdir(parents=True, exist_ok=True)

MY_PROFILE = {
    "currentRate": 5.87575757575,
    "loanType": "5/1 ARM",
    "location": "Milpitas, CA 95035",
    "propertyType": "Single-family detached",
}

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
}


# Shared HTTP client with connection pooling
_http_client = None


def get_client():
    global _http_client
    if _http_client is None:
        _http_client = Client(
            headers=HEADERS,
            timeout=12.0,
            follow_redirects=True,
            limits=httpx.Limits(max_connections=10, max_keepalive_connections=5),
        )
    return _http_client


def fetch_html(url: str, timeout: float = 12.0) -> str:
    client = get_client()
    r = client.get(url, timeout=timeout)
    r.raise_for_status()
    return r.text


# ── Supabase upload ──────────────────────────────────────────────


def upload_to_supabase(result: dict) -> bool:
    """Push today's rates to Supabase mortgage_rates table for historical tracking."""
    supabase_url = os.environ.get("SUPABASE_URL")
    supabase_key = os.environ.get("SUPABASE_SERVICE_KEY")
    if not supabase_url or not supabase_key:
        print("  ⚠ Supabase: SUPABASE_URL/SUPABASE_SERVICE_KEY not set, skipping")
        return False

    institutions = result.get("institutions", {})
    rows = []
    today = result["date"]

    # Pick lowest rate per (lender, term) — the DB unique constraint is on (date, lender, term)
    best = {}
    for _, inst in institutions.items():
        name = inst.get("name", "Unknown")
        for entry in inst.get("fixed", []) + inst.get("arm", []):
            term = entry["term"]
            rate = entry["rate"]
            key = (name, term)
            if key not in best or rate < best[key]["rate"]:
                best[key] = {
                    "date": today,
                    "lender": name,
                    "term": term,
                    "rate": rate,
                    "apr": entry.get("apr"),
                }
    rows = list(best.values())

    if not rows:
        print("  ⚠ Supabase: no rows to upload")
        return False

    # Upsert in batches of 50
    import urllib.request
    import ssl

    ctx = ssl.create_default_context()
    headers = {
        "apikey": supabase_key,
        "Authorization": f"Bearer {supabase_key}",
        "Content-Type": "application/json",
        "Prefer": "resolution=merge-duplicates",
    }

    success_count = 0
    batch_size = 50
    for i in range(0, len(rows), batch_size):
        batch = rows[i:i + batch_size]
        data = json.dumps(batch).encode("utf-8")
        # Upsert: POST + on_conflict query param tells PG which columns form the conflict target
        req = urllib.request.Request(
            f"{supabase_url}/rest/v1/mortgage_rates?on_conflict=date,lender,term",
            data=data,
            headers=headers,
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=15, context=ctx) as resp:
                status = resp.status
                if status in (200, 201):
                    success_count += len(batch)
                else:
                    print(f"  ⚠ Supabase batch {i//batch_size}: HTTP {status}")
        except Exception as e:
            print(f"  ⚠ Supabase batch {i//batch_size} failed: {e}")

    print(f"  ✓ Uploaded {success_count}/{len(rows)} rate rows to Supabase")
    return True


# ── Scrapers ──────────────────────────────────────────────────────


def run_all_scrapers():
    from scrapers.patelco import scrape as sc_patelco
    from scrapers.starone import scrape as sc_starone
    from scrapers.wellsfargo import scrape as sc_wf
    from scrapers.golden1 import scrape as sc_golden1
    from scrapers.bofa import scrape as sc_bofa
    from scrapers.chase import scrape as sc_chase
    from scrapers.citi import scrape as sc_citi
    from scrapers.pmms import scrape as sc_pmms

    scrapers = [
        ("patelco", sc_patelco),
        ("starone", sc_starone),
        ("wellsfargo", sc_wf),
        ("golden1", sc_golden1),
        ("bofa", sc_bofa),
        ("chase", sc_chase),
        ("citi", sc_citi),
    ]

    institutions = {}
    errors = []

    for name, scraper in scrapers:
        try:
            print(f"→ Fetching {name}...")
            result = scraper(fetch_html)
            institutions[name] = result
            fixed_count = len(result.get("fixed", []))
            arm_count = len(result.get("arm", []))
            print(f"  ✓ {result['name']}: {fixed_count} fixed, {arm_count} ARM rates")
        except Exception as e:
            print(f"  ✗ {name} failed: {e}")
            errors.append({"source": name, "error": str(e)})

    # PMMS (national averages)
    print("→ Fetching pmms (Freddie Mac national averages)...")
    try:
        pmms_result = sc_pmms(fetch_html)
        institutions["pmms"] = pmms_result
        print(f"  ✓ PMMS: {pmms_result.get('rates', {})}")
    except Exception as e:
        print(f"  ✗ pmms failed: {e}")
        errors.append({"source": "pmms", "error": str(e)})

    return institutions, errors


# ── Summary ────────────────────────────────────────────────────────


def build_summary(institutions: dict) -> dict:
    all_fixed_30 = []
    all_fixed_15 = []
    all_arm_5_1 = []
    all_arm_7_1 = []
    all_arm_10_1 = []

    for inst in institutions.values():
        for f in inst.get("fixed", []):
            term = f.get("term", "").lower()
            if "30" in term:
                all_fixed_30.append(f["rate"])
            elif "15" in term:
                all_fixed_15.append(f["rate"])
        for a in inst.get("arm", []):
            term = a.get("term", "").lower()
            if "5/1" in term or "5-1" in term:
                all_arm_5_1.append(a["rate"])
            elif "7/1" in term or "7-1" in term or "7/6" in term:
                all_arm_7_1.append(a["rate"])
            elif "10/1" in term:
                all_arm_10_1.append(a["rate"])

    pmms = institutions.get("pmms", {})
    pmms_rates = pmms.get("rates", {})

    def avg(lst):
        return round(sum(lst) / len(lst), 3) if lst else None

    def best(lst):
        return round(min(lst), 3) if lst else None

    summary = {
        "30_year_fixed_avg": avg(all_fixed_30),
        "30_year_fixed_best": best(all_fixed_30),
        "15_year_fixed_avg": avg(all_fixed_15),
        "15_year_fixed_best": best(all_fixed_15),
        "5_1_arm_avg": avg(all_arm_5_1),
        "5_1_arm_best": best(all_arm_5_1),
        "7_1_arm_avg": avg(all_arm_7_1),
        "7_1_arm_best": best(all_arm_7_1),
        "10_1_arm_avg": avg(all_arm_10_1),
        "national_30_year_fixed": pmms_rates.get("30_year_fixed"),
        "national_15_year_fixed": pmms_rates.get("15_year_fixed"),
        "my_rate": MY_PROFILE["currentRate"],
        "my_loan_type": MY_PROFILE["loanType"],
        "institutions_count": len(institutions),
    }

    summary["30_year_fixed"] = summary["30_year_fixed_avg"] or pmms_rates.get("30_year_fixed")
    summary["15_year_fixed"] = summary["15_year_fixed_avg"] or pmms_rates.get("15_year_fixed")
    summary["5_1_arm"] = summary["5_1_arm_avg"]
    summary["7_1_arm"] = summary["7_1_arm_avg"]
    summary["best_7_1_arm"] = summary["7_1_arm_best"]
    summary["10_1_arm"] = summary["10_1_arm_avg"]
    summary["best_30yr_lender_rate"] = summary["30_year_fixed_best"]

    market_arm = summary["5_1_arm_avg"] or pmms_rates.get("30_year_fixed")
    if market_arm:
        diff = round(market_arm - MY_PROFILE["currentRate"], 3)
        if diff > 0.5:
            assessment = "🌟 Your rate is EXCELLENT — well below market"
        elif diff > 0.25:
            assessment = "✅ Your rate is BETTER than market"
        elif diff > 0.05:
            assessment = "🟢 Your rate is slightly better than market"
        elif diff > -0.05:
            assessment = "🟡 Market is about equal to your rate"
        elif diff > -0.25:
            assessment = "🟠 Market is slightly better — watch for refinance"
        else:
            assessment = "🔴 Market is BETTER — consider refinancing"

        summary["vs_my_rate"] = {
            "market_rate": market_arm,
            "my_rate": MY_PROFILE["currentRate"],
            "difference": diff,
            "assessment": assessment,
        }

    return summary


# ── Main ──────────────────────────────────────────────────────────


def main():
    now = datetime.now(timezone.utc)
    today = now.strftime("%Y-%m-%d")
    print(f"[{now.isoformat()}] Fetching mortgage rates for {today}...")

    institutions, errors = run_all_scrapers()
    summary = build_summary(institutions)

    result = {
        "date": today,
        "fetchedAt": now.isoformat(),
        "profile": MY_PROFILE,
        "institutions": institutions,
        "summary": summary,
    }

    if errors:
        result["errors"] = errors

    # Save local files
    day_file = DATA_DIR / f"rates-{today}.json"
    day_file.write_text(json.dumps(result, indent=2))
    print(f"\nSaved: {day_file}")

    latest_file = SITE_DATA_DIR / "latest.json"
    latest_file.write_text(json.dumps(result, indent=2))
    print(f"Saved: {latest_file}")

    # Upload to Supabase for historical tracking
    print("\n→ Uploading to Supabase...")
    upload_to_supabase(result)

    # Dashboard
    print("\n" + "=" * 60)
    print(f"📊 MORTGAGE RATE DASHBOARD — {today}")
    print("=" * 60)

    s = summary
    fmt = lambda v: f"{v:.3f}%" if v else "N/A"
    print(f"30-Year Fixed:  avg={fmt(s['30_year_fixed_avg'])}  best={fmt(s['30_year_fixed_best'])}")
    print(f"15-Year Fixed:  avg={fmt(s['15_year_fixed_avg'])}  best={fmt(s['15_year_fixed_best'])}")
    print(f"5/1 ARM:        avg={fmt(s['5_1_arm_avg'])}  best={fmt(s['5_1_arm_best'])}")
    print("-" * 60)
    print(f"Your Rate (5/1 ARM): {MY_PROFILE['currentRate']}%")
    if "vs_my_rate" in s:
        v = s["vs_my_rate"]
        print(f"Market rate:            {v['market_rate']:.3f}%")
        print(f"Difference:             {v['difference']:+.3f}%")
        print(f"Status:                 {v['assessment']}")

    print(f"\nSources: {len(institutions)} institutions")
    for name, inst in institutions.items():
        fixed = len(inst.get("fixed", []))
        arm = len(inst.get("arm", []))
        print(f"  {inst['name']}: {fixed} fixed, {arm} ARM")

    if errors:
        print(f"\n⚠️ {len(errors)} sources failed:")
        for e in errors:
            print(f"  {e['source']}: {e['error']}")

    print("=" * 60)
    return result


if __name__ == "__main__":
    sys.path.insert(0, str(ROOT))
    main()